### PR TITLE
doc: document various best practices and expectations

### LINKED ISSUE #11976: APIs that support a callback should provide both the device pointer and a generic pointer
I propose both a pointer to the associated device and a generic pointer should be provided consistently for all device API callbacks.

I'm running into an issue right now with `gpio_callback` where what comes in is the GPIO device `port` which I need to change the interrupt level sense, a pointer `cb` to a callback structure that lives within a sensor's driver data structure, and a bit-mask of pins associated with the callback.

From that I can get the sensor's `dev->driver_data` pointer by:
```
	struct sht3xd_data *drv_data =
		CONTAINER_OF(cb, struct sht3xd_data, gpio_cb);
```

But I can't get the sensor's `dev->config->config_info` pointer which has the data that tells me which pin is which, because the base device pointer to the sensor device has been erased.  (Yes, I may share a callback with multiple pins: the API permits/encourages this by providing a mask rather than a pin.)

As a workaround I'm going to have to store a copy of the device pointer in the `sht3xd_data` structure.  Had the `gpio_callback_handler_t` taken a generic pointer instead of a pointer to the callback, I could have passed it the sensor device pointer from which I could get both the callback structure and the configuration pointer.  People without this need could have passed the pointer to the callback structure, and have what they have today.

A callback for a device should always be given the device for which it's a callback for consistency across APIs where this usually necessary, and for future-proofing for API enhancements where it may become necessary.  The API design should not assume that the device pointer can be fetched by doing offset manipulations based on a user-provided pointer.

Further, callbacks should always be passed a user-provided generic pointer rather than a pointer to the callback object, for the same reason of allowing flexibility and not making developers try to reconstruct it from a pointer to an embedded member.

### PR DESCRIPTION
**Pull request by pabigot**
This records the consensus from some recent API telecon discussions related to devicetree hardware vs software configuration, callback design, and how to ensure documentation is generated for conditional capabilities.

Closes #11976.

### CODE REVIEW

**Comment by pabigot** [comment]
@henrikbrixandersen We discussed this in API on 2020-07-28 where there was some issue that both you and I were interested in, and I took the action to address documenting it but didn't record what it was.  I'm pretty sure it was related to documenting conditional code.  If you remember it being something else, please remind me.

File: doc/guides/design_guidelines.rst [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+* The next parameter(s) should be additional information specific to the
+  callback invocation, such as a channel identifier, new status value,
+  and/or a message pointer followed by the message length.
+
+* The final parameter should be a ``void *user_data`` pointer carrying
+  context that allows a shared callback function to locate additional
+  material necessary to process the callback.
+
+An exception to providing ``user_data`` as the last parameter may be
```
Comment (nashif) [inline doc/guides/design_guidelines.rst:31]: would be great if you added a few examples to illustrate that with some existing or a dummy API...

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen We discussed this in API on 2020-07-28 where there was some issue that both you and I were interested in, and I took the action to address documenting it but didn't record what it was. I'm pretty sure it was related to documenting conditional code. If you remember it being something else, please remind me.

@pabigot Yes, I believe it was documentation of conditional code - and which "parts" of a conditional API to hide behind `#ifdef` (API functions, API struct member functions, userspace handlers, ...).

File: doc/guides/design_guidelines.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+In the case where a feature is determined to be optional the following
+practices should be followed.
+
+First, any data that is accessed only when the feature is enabled should
+be conditionally included via ``ifdef CONFIG_FOO`` in the structure or
+union declaration.  This reduces memory use for applications that don't
+need the capability.
+
+Second, function declarations that are available only when the option is
+available may be handled in the either of these ways:
```
Comment (henrikbrixandersen) [inline doc/guides/design_guidelines.rst:59]: Why have two recommended practices for this? How does one choose the best one?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, thanks for doing this. Just one small question/nit, that I would like cleared up:

File: doc/guides/design_guidelines.rst (hunk 2 above)
Comment (pabigot) [reply doc/guides/design_guidelines.rst:59]: Both approaches are present in existing headers.  The rationale for preferring one over another is in the following paragraph, but there is no strong recommendation.  

I'm ambivalent about having both.  Removing the first (unconditional declarations+documentation) would be OK and maybe that's what should happen.  Then there should be a task to go through the headers and remove the unnecessary exclusions, and update the per-function documentation to explicitly note the dependency, to make the codebase consistent with the recommendations.

Comment (henrikbrixandersen) [reply doc/guides/design_guidelines.rst:59]: I think we should try to pick only one (and I agree it should be unconditional declarations and documentation). Any APIs currently in violation of this should then be updated (can be done in separate PRs) to match the new recommendation.

Comment (pabigot) [reply doc/guides/design_guidelines.rst:59]: OK, we'll go that way.

**Review by henrikbrixandersen — APPROVED** [review]

File: doc/guides/design_guidelines.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  conditionally included via ``#ifdef CONFIG_MYFEATURE`` in the
+  structure or union declaration.  This reduces memory use for
+  applications that don't need the capability.
+* Function declarations that are available only when the option is
+  enabled should be provided unconditionally.  Add a note in the
+  description that the function is available only when the specified
+  feature is enabled, referencing the required Kconfig symbol by name.
+  (In cases where the function is used but not enabled a link-time error
+  will be generated.)
+* Function definitions in source code modules that provide unconditional
```
Comment (chrta) [inline doc/guides/design_guidelines.rst:92]: `that provide` -> `that also provide`
I had to read the following bullet point to get the meaning of this one. I think adding the "also" or sth like that clarifies the meaning here.

Another option would be to move this item below the "* Source code modules that provide only conditionally available code..."

**Review by chrta — CHANGES_REQUESTED** [review]
This is a good addition, just one minor nit.

File: doc/guides/design_guidelines.rst (hunk 1 above)
Comment (pabigot) [reply doc/guides/design_guidelines.rst:31]: Done, below

File: doc/guides/design_guidelines.rst (hunk 3 above)
Comment (pabigot) [reply doc/guides/design_guidelines.rst]: Done

**Review by chrta — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
Looks great, thanks @pabigot

**Pull request merged**